File: app/src/main/java/com/exzell/mangaplayground/reader/ReaderPager.java
package com.exzell.mangaplayground.reader;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.view.GestureDetectorCompat;
import androidx.viewpager.widget.ViewPager;

import java.util.function.Consumer;
import java.util.function.Function;

import kotlin.jvm.functions.Function1;


public class ReaderPager extends ViewPager {

    private Function1<MotionEvent, Boolean> onSingleTap;

    private Consumer<Integer> onPageChanged;

    private GestureDetectorCompat mGestureDetector = new GestureDetectorCompat(getContext(), new SwipeListener(null, new Function1<MotionEvent, Boolean>() {
        @Override
        public Boolean invoke(MotionEvent motionEvent) {
            return onSingleTap.invoke(motionEvent);
        }
    }));

    private SimpleOnPageChangeListener mPageListener = new SimpleOnPageChangeListener(){
        @Override
        public void onPageSelected(int position) {
            onPageChanged.accept(position);
        }
    };


    public ReaderPager(@NonNull Context context) {
        super(context);
        init();
    }

    public ReaderPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init(){
        addOnPageChangeListener(mPageListener);
    }

    public void setOnSingleTap(Function1<MotionEvent, Boolean> onSingleTap) {
        this.onSingleTap = onSingleTap;
    }

    public void setOnPageChanged(Consumer<Integer> onPageChanged) {
        this.onPageChanged = onPageChanged;
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        boolean ret = mGestureDetector.onTouchEvent(ev);
        if(!ret) ret = super.dispatchTouchEvent(ev);
        return ret;
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        clearOnPageChangeListeners();
    }
}

File: app/src/main/java/com/exzell/mangaplayground/reader/ReaderPager.kt
package com.exzell.mangaplayground.reader

import android.content.Context
import android.util.AttributeSet
import android.view.GestureDetector
import androidx.viewpager.widget.ViewPager
import android.view.MotionEvent
import androidx.core.view.GestureDetectorCompat
import timber.log.Timber

class ReaderPager(context: Context, attrs: AttributeSet?) : ViewPager(context, attrs) {


    var onSingleTap: ((MotionEvent) -> Boolean)? = null
    var onPageChanged: ((Int) -> Unit)? = null
    private val mGestureDetector: GestureDetectorCompat

    private val mPageListener: SimpleOnPageChangeListener = object : SimpleOnPageChangeListener() {
        override fun onPageSelected(position: Int) {
            onPageChanged?.invoke(position)
        }
    }

    constructor(context: Context) : this(context, null)

    init {
        addOnPageChangeListener(mPageListener)
        mGestureDetector = GestureDetectorCompat(context, createGestureListener())
    }

    private fun createGestureListener() = object : GestureDetector.SimpleOnGestureListener() {
        override fun onDown(e: MotionEvent?) = true

        override fun onSingleTapConfirmed(e: MotionEvent?) = onSingleTap?.invoke(e!!) ?: false
    }

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
        Timber.d("Gesture detector reached with event $ev")
        super.onInterceptTouchEvent(ev)
        return true
    }

    override fun onTouchEvent(ev: MotionEvent?): Boolean {
        mGestureDetector.onTouchEvent(ev)
        return super.onTouchEvent(ev)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        clearOnPageChangeListeners()
    }
}